Step2

bird798 opened the pull request:
안녕하세요.

2단계 문자열 사칙 연산 계산기 구현 리뷰요청 드립니다.

감사합니다.

chan798 committed:
Test

chan798 committed:
문자열 사칙 연산 계산기 구현

chan798 committed:
불필요한 파일 삭제

catsbi commented on src/main/kotlin/step2/Calculator.kt:
```diff
@@ -0,0 +1,71 @@
+class Calculator() {
+    fun calculate(formula: String): Double {
+        val factors = split(formula)
+        // 리스트가 비어 있거나, Factor가 홀수가 아니거나.
```
꼭 필요한 주석이 아니라면 제거하는걸 권장드려요. 
주석이 최소화되야 할 여러 이유가 있는데, 다음과 같으니 참고해보면 좋을 것 같아요. 

- 주석도 코드와 커플링이 발생한다. 코드가 바뀔경우 주석도 바뀌어야 하는데  컴파일단계에서 IDE에 의해 경고를 줄 수 없기 때문에 주석 업데이트를 놓칠 경우 주석의 내용과 함수의 실제 동작이 달라서 유지보수성을 떨어트릴 수 있다. 
- 함수/변수/클래스명을 충분히 의미있게 잘 지었다면 주석이 없어도 이해하는데 문제가 없을 것이다. 그게 안되어 주석을 작성한다면 네이밍을 잘 못한게 아닐까 고민하자. 
- 첫 번째 이유와 연결하여 불필요한 주석이 많을 경우 주석에 대한 집중도가 떨어져 정말 중요한 주석도 보지 않게되어 문제를 야기시킬 수 있다.

bird798 replied:
불필요한 주석 제거하였습니다.

catsbi commented on src/main/kotlin/step2/Calculator.kt:
```diff
@@ -0,0 +1,71 @@
+        var result = 0.0
+        var operation = ""
+        for (i in 0 until factors.size step 1) {
+            if ((i % 2) == 0) {
```
반복문안에 분기문이 있다보니 코드의 indent가 깊어지는 것 같아요 인덴트를 하나 더 줄여볼 수 없을까요?
그리고 `i % 2 == 0`이런 코드는 함수로 분리해서 isEven(i) 이런식으로 나누면 가독성이 더 높아질 수 있지 않을까요?

bird798 replied:
루프 안에 조건문을 제거하였습니다.

catsbi commented on src/main/kotlin/step2/Calculator.kt:
```diff
@@ -0,0 +1,71 @@
+                    "" -> result = number // 저장된 연산자가 없으면 현재것 저장
+                    else -> throw IllegalArgumentException("정의되지 않은 연산자입니다.")
+                }
+                operation = ""
```
추가되는 연산자가 생길경우 코드가 계속 수정되야할 것 같아요. 연산자가 정해져있고, 각 연산자별 행위를 각각의 인스턴스에게 위임할 수 있도록 열거타입을 활용해보는건 어떨까요?

bird798 replied:
열거형으로 변경하였고, 열거형 안애 함수를 넣어 획득된 열거형 개체에 포함된 Function으로 When문을 대체하였습니다.

catsbi commented on src/main/kotlin/step2/Calculator.kt:
```diff
@@ -0,0 +1,71 @@
+
+        var result = 0.0
+        var operation = ""
+        for (i in 0 until factors.size step 1) {
```
반복문 내의 로직을 잘 보면 연산자와 피연산자를 추출하고 있는 것 같아요. 
행위들이 여러개가 있다보니 분기문이들어가고 when절이 들어가고 있으면서 하나의 로직에 여러 책임들이 맞물리는 것 같습니다. 

이럴 경우 테스트도 힘들것 같아요. 테스트의 브랜치 커버리지도 너무 복잡해서 높히기 힘들어보이구요. 

- 연산자 추출및 유효성 검증
- 피연산자 추출및 유혀성검증
- 계산

이정도로 행위를 나눠서 해보는건 어떨까요?

bird798 replied:
분기문과 When절을 제거하였고, 
루프 안에서 아래의 동작을 단순화하였습니다.

* 연산자 추출및 유효성 검증
* 피연산자 추출및 유혀성검증
* 계산

catsbi commented on src/main/kotlin/step2/Calculator.kt:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    private fun split(formula: String): List<String> {
+        val factors = formula.trim().split(" ")
```
" "는 구분자로 사용되는것 같은데 companion object 블럭으로 상수 분리해서 가독성을 높혀보면 어떨까요? 

매직리터럴이나 매직넘버는 그자체로 가독성을 주기 힘든 경우가 많고 재사용성 이슈도 있어 상수로 분리하며 네이밍을 통해 가독성을 높히고 재사용성도 챙길 수 있습니다. 

물론, 재사용성이 필요없고 값 자체로 의미를 제공할 수 있을 경우엔 고려할 필요가 있지만요.

bird798 replied:
" "를 상수로 변경하였습니다.

catsbi commented on src/test/kotlin/step2/CalculatorTest.kt:
```diff
@@ -0,0 +1,28 @@
+        assertThat(calculator.calculate("1 * 2")).isEqualTo(2.0)
+        assertThat(calculator.calculate("1 / 2")).isEqualTo(0.5)
+        assertThat(calculator.calculate("2 + 3 * 4 / 2")).isEqualTo(10.0)
+    }
```
ParameterizedTest를 통해 단언문들 중복을 제거해보면 어떨까요?

bird798 replied:
ParameterizedTest를 이용해 단언문을 제거하였습니다.

catsbi commented on src/test/kotlin/step2/CalculatorTest.kt:
```diff
@@ -0,0 +1,28 @@
+
+    @ParameterizedTest
+    @ValueSource(strings = ["=", "1 a 1", "1 -", "- 1", "1+1", "1+ 1"])
+    fun testCalulateException(input: String?) {
```
무슨 테스트를 하는지 코드를 봐야만 알 수 있다면 테스트가 늘어날수록 테스트 분석에 대한 피로도가 높아질 것 같아요. 
DisplayName등을 통해 어떤 함수가 어떤 조건에서 어떤 결과가 유추되는지에 대해 작성해서 가독성을 높히면 어떨까요?

bird798 replied:
DisplayName 어노테이션을 통해 TestCase에 한글 설명을 추가하였습니다.

catsbi reviewed (changes requested):
찬희님 2단계 미션 잘 구현해주셨어요 👍 

몇 가지 고민해보면 좋을 부분들이 있어 코멘트 남겨드렸으니 확인 후 다시 리뷰 요청 부탁드려요!

chan798 committed:
* Testcase
- DisplayName추가.
- 반복되는 단언문 삭제

* Calculator
- 열거형에 함수를 대입하여 When문을 대체
- 구분문자(" ")를 상수로 정의
- 불필요한 주석 제거

chan798 committed:
1. 변수명 변경
2. Error Message 상수값으로 변경
3. 중복된 에러 처리 삭제

catsbi commented on src/main/kotlin/step2/Calculator.kt:
```diff
@@ -1,71 +1,48 @@
+        PLUS("+", { a, b -> a + b }),
+        MINUS("-", { a, b -> a - b }),
+        MULTIPLY("*", { a, b -> a * b }),
+        DIVIDE("/", { a, b -> a / b }),
```
나눗셈은 0으로 나눗셈할 경우 ArithmeticException이 발생할 것 같은데 사전에 유효성 검증 로직을 추가해주는건 어떨까요? 

모든 예외는 최대한 빠르게 발견하는게 좋으니까요 ㅎㅎ

bird798 replied:
네 다음 부분에 고려하겠습니다.

catsbi commented on src/main/kotlin/step2/Calculator.kt:
```diff
@@ -1,71 +1,48 @@
+        ;
+
+        companion object {
+            private val map = entries.associateBy(OPERATOR::operator)
+
+            fun get(operation: String): OPERATOR? {
+                return map[operation]
```
캐싱처리 💯 

get함수에서 operation이 정상적이지 않을 경우에 대한 유효성검증은 필요 없을까요?!

bird798 replied:
1. 커멘트 주신 "캐싱처리 100" 은 어떤 의미실까요?

2. get함수에 잘못된 값이 들어왔을 경우 Map에 해당 사항이 없으니, Null이 반환되지 않을까요?
그럼 아래의 코드에서 걸러졌을 꺼 같은데요.

```
val operator = OPERATOR.get(factors[i]) ?: throw IllegalArgumentException(ERROR_MESSAGE_ILLEGAL_OPERATOR)

```

catsbi reviewed (approved):
찬희님 피드백 반영 잘 해주셨어요 👍 

코드 인덴트가 줄어들고, 가독성도 높아진 것 같아요. 
2단계 머지하도록 하겠습니다. 본격적인 3단계 미션 진행해주세요!
